06_puppy/hsy3418/syncStore.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+)
+
+type SyncStore struct {
+	sync.Mutex
+	sync.Map
+}
+
+func NewSyncStore() *SyncStore {
+	return &SyncStore{}
+}
+
+// CreatePuppy adds a nuw puppy to the puppies store
+func (m *SyncStore) CreatePuppy(puppy Puppy) error {
+	m.Lock()
+	defer m.Unlock()
+	if _, exists := m.Load(puppy.ID); exists {
+		return fmt.Errorf("puppy with %d ID already exists", puppy.ID)
+	}
+	m.Store(puppy.ID, puppy)
+	return nil
+}
+
+// ReadPuppy retrieves the puppy for a given id from puppies store
+func (m *SyncStore) ReadPuppy(id int32) (Puppy, error) {
+	if p, exists := m.Load(id); exists {
```

Comment:
ok instead of exists and invert logic so that error is inside if: if !ok return err

Reply:
I need the p within the if block, so a puppy could be created if ok, otherwise I need to have an else block for create the puppy if ok, but if doing so, I couldn't pass lint..So it seems like I need to keep this logic.

Reply:
If you flip the logic, you won't need the p inside the if statement, and can go about it that way. 

```go
func (m *SyncStore) ReadPuppy(id int32) (Puppy, error) {
  p, ok := m.Load(id)
  if !ok {
    return Puppy{}, fmt.Errorf("puppy with %d ID does not exist", id)
  }
  puppy, _ := p.(Puppy)
  return puppy, nil
}
```

I think what Julia is suggesting here is to have your structure following the convention of: **if err -> fail early**

Normally if there are multiple error cases, I think this is a good approach. When there is a single error case, I am not as fussed either way. I will leave this one up to you.

Reply:
Got it!